presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -266,26 +266,33 @@ private static PlanNodeStatsEstimate filterStatsForUnknownExpression(PlanNodeSta
                     .collect(ImmutableList.toImmutableList());
 
             if (!valuesEqualityStats.stream().allMatch(Optional::isPresent)) {
-                return visitExpression(node, context);
+                return Optional.empty();
             }
 
             PlanNodeStatsEstimate statsSum = valuesEqualityStats.stream()
                     .map(Optional::get)
                     .reduce(filterForFalseExpression().get(), PlanNodeStatsEstimateMath::addStatsAndSumDistinctValues);
 
             if (isNaN(statsSum.getOutputRowCount())) {
-                return visitExpression(node, context);
+                return Optional.empty();
             }
 
-            Symbol inValueSymbol = Symbol.from(node.getValue());
-            SymbolStatsEstimate symbolStats = input.getSymbolStatistics(inValueSymbol);
-            double notNullValuesBeforeIn = input.getOutputRowCount() * (1 - symbolStats.getNullsFraction());
+            Optional<Symbol> inValueSymbol = asSymbol(node.getValue());
+            SymbolStatsEstimate inValueStats = getExpressionStats(node.getValue());
+            if (Objects.equals(inValueStats, UNKNOWN_STATS)) {
+                return Optional.empty();
+            }
 
-            SymbolStatsEstimate newSymbolStats = statsSum.getSymbolStatistics(inValueSymbol)
-                    .mapDistinctValuesCount(newDistinctValuesCount -> min(newDistinctValuesCount, symbolStats.getDistinctValuesCount()));
+            double notNullValuesBeforeIn = input.getOutputRowCount() * (1 - inValueStats.getNullsFraction());
 
-            return Optional.of(input.mapOutputRowCount(rowCount -> min(statsSum.getOutputRowCount(), notNullValuesBeforeIn))
-                    .mapSymbolColumnStatistics(inValueSymbol, oldSymbolStats -> newSymbolStats));
+            PlanNodeStatsEstimate estimate = input.mapOutputRowCount(rowCount -> min(statsSum.getOutputRowCount(), notNullValuesBeforeIn));
+
+            if (inValueSymbol.isPresent()) {
+                SymbolStatsEstimate newSymbolStats = statsSum.getSymbolStatistics(inValueSymbol.get())
+                        .mapDistinctValuesCount(newDistinctValuesCount -> min(newDistinctValuesCount, inValueStats.getDistinctValuesCount()));
+                estimate = estimate.mapSymbolColumnStatistics(inValueSymbol.get(), oldSymbolStats -> newSymbolStats);
```

Comment:
I would just build it with a builder instead

Reply:
I left the existing wording.

Reply:
No change is required. Feel free to leave it as is.

I just looks like a yet another lambda abuse.

Reply:
i sometimes had an impression that we shouldn't have these `map*` (`mapSymbolColumnStatistics`, `mapDistinctValuesCount`, ..) methods at all. We should approach this holistically though -- review use cases, see if it would be better not to have these methods....  However -- slightly later, to avoid conflicts.